src/V3MergeCond.cpp

```diff
@@ -51,71 +79,363 @@
+bool areDisjoint(const std::set<const AstVar*>& a, const std::set<const AstVar*>& b) {
+    if (a.empty() || b.empty()) return true;
+    const auto endA = a.end();
+    const auto endB = b.end();
+    auto itA = a.begin();
+    auto itB = b.begin();
+    while (true) {
+        if (*itA == *itB) return false;
+        if (std::less<const AstVar*>{}(*itA, *itB)) {
+            itA = std::lower_bound(++itA, endA, *itB);
+            if (itA == endA) return true;
+        } else {
+            itB = std::lower_bound(++itB, endB, *itA);
+            if (itB == endB) return true;
+        }
+    }
+}
+
+//######################################################################
+// Structure containing information required for code motion/merging
+
+struct StmtProperties {
+    AstNode* m_condp = nullptr;  // The condition expression, if a conditional node
+    std::set<const AstVar*> m_rdVars;  // Variables read by this statement
+    std::set<const AstVar*> m_wrVars;  // Variables writen by this statement
+    bool m_isFence = false;  // Nothing should move across this statement, nor should it be merged
+    AstNodeStmt* m_prevWithSameCondp = nullptr;  // Previous node in same list, with same condition
+    bool writesConditionVar() const {
+        // This relies on MarkVarsVisitor having been called on the condition node
+        for (const AstVar* const varp : m_wrVars) {
+            if (varp->user1()) return true;
+        }
+        return false;
+    }
 };
 
-class CheckMergeableVisitor final : public VNVisitor {
-private:
-    // STATE
-    bool m_condAssign = false;  // Does this tree contain an assignment to a condition variable??
-    bool m_impure = false;  // Does this tree contain an impure node?
+// We store the statement properties in user3 via AstUser3Allocator
+using StmtPropertiesAllocator = AstUser3Allocator<AstNodeStmt, StmtProperties>;
 
-    // METHODS
-    VL_DEBUG_FUNC;  // Declare debug()
+//######################################################################
+// Code motion analysis and implementation
 
-    // VISITORS
-    virtual void visit(AstNode* nodep) override {
-        if (m_impure) return;
-        // Clear if node is impure
-        if (!nodep->isPure()) {
-            UINFO(9, "Not mergeable due to impure node" << nodep << endl);
-            m_impure = true;
-            return;
+// Pure analysis visitor that build the StmtProperties for each statement in the given
+// AstNode list (following AstNode::nextp())
+class CodeMotionAnalysisVisitor final : public VNVisitor {
+    // NODE STATE
+    // AstNodeStmt::user3   -> StmtProperties (accessed via m_stmtProperties, managed externally,
+    //                         see MergeCondVisitor::process)
+    // AstNode::user4       -> Used by V3Hasher
+    // AstNode::user5       -> Points from condition node to last conditional with that
+    //                         condition so far encountered in the same AstNode list
```

Comment:
```suggestion
    // AstNode::user5       -> AstNode*.  Points from condition node to last conditional with that
    //                         condition so far encountered in the same AstNode list
```

Please update these sort of comments to match style else where with type stored in the user4 first.  Likewise many "Flag {etc}" should be "bool.  {etc}"